### PR TITLE
[experiment] Reference arities

### PR DESCRIPTION
**Pull request by mattam82**
Add an optional minimal arity information to constants, that can be used to avoid spurious unfoldings during unifications.
The goal of this experiment is to see if this can have a measurable impact on the (mis)use of definitional classes in math-classes. Example use: `#[arity="2"] Definition equiv {A} {Equiv : Equiv A} := Equiv.` disables unfolding of partial applications of `equiv`: it unfolds only when applied to 2 or more arguments. This avoids unification going through useless paths due to the first-order approximation heuristic, where it will try to unify the body `fun (A : Type@{i}) (Equiv : Equiv A) => Equiv` with other terms not headed by `equiv` and separately its arguments.

This also supports forcing the eta-expansion of constructors at elaboration time, but is disabled for now as this introduces incompatibilities. In particular, when syntactically matching for `S` in `change` calls, `S` is pretyped and expanded to `fun x => S x` which must now match a lambda-abstraction. This is actually the only incompatibility found in the stdlib and test-suite when forcing fully-applied constructors. 

For inductive applications, we do force template-polymorphic inductives to be applied to at least their arguments whose type mention universes, and refresh the arguments during expansion, so as to avoid capture of global template universes. This might be better handled by @ppedrot and @SkySkimmer 's work on reinterpreting template-polymorphism as sort-poly+implicit univ-poly. Like for constructors, it introduces syntactic incompatibilities. E.g. `Monad option` becomes `Monad (fun H : Type@{fresh_i} => option H)` which avoids capture but also introduces an eta-expansion which can impact code doing syntactic matching. Note that unification does not respect the arity yet when instantiating evars/metas.

### CODE REVIEW

**Comment by mattam82** [comment]
@coqbot bench

File: dev/ci/user-overlays/19132-mattam82-reference-arities.sh [hunk 1]
```diff
@@ -0,0 +1 @@
+overlay math_classes https://github.com/mattam82/math-classes reference-arities 19132
```
Comment (SkySkimmer) [inline dev/ci/user-overlays/19132-mattam82-reference-arities.sh:1]: note that the bench doesn't use overlays, you need something like https://github.com/coq/coq/pull/19131/files if you want to use this branch of mathclasses in the bench

Comment (mattam82) [reply dev/ci/user-overlays/19132-mattam82-reference-arities.sh:1]: Good to know, thanks!

File: dev/bench/bench.sh [hunk 2]
```diff
@@ -70,15 +70,15 @@ check_variable () {
 : "${old_ocaml_flambda:=0}"
 : "${new_coq_repository:=$CI_REPOSITORY_URL}"
 : "${old_coq_repository:=$CI_REPOSITORY_URL}"
-: "${new_coq_opam_archive_git_uri:=https://github.com/coq/opam-coq-archive.git}"
+: "${new_coq_opam_archive_git_uri:=https://github.com/mattam82/opam-coq-archive.git}"
```
Comment (SkySkimmer) [inline dev/bench/bench.sh:73]: no such repo found, is it private?

Comment (mattam82) [reply dev/bench/bench.sh:73]: No, it's the usual opam-coq-archive vs coq-opam-archive confusion :)

**Comment by mattam82** [comment]
@coqbot bench

File: dev/bench/bench.sh [hunk 3]
```diff
@@ -70,15 +70,15 @@ check_variable () {
 : "${old_ocaml_flambda:=0}"
 : "${new_coq_repository:=$CI_REPOSITORY_URL}"
 : "${old_coq_repository:=$CI_REPOSITORY_URL}"
-: "${new_coq_opam_archive_git_uri:=https://github.com/coq/opam-coq-archive.git}"
+: "${new_coq_opam_archive_git_uri:=https://github.com/mattam82/coq-opam-archive.git}"
 : "${old_coq_opam_archive_git_uri:=https://github.com/coq/opam-coq-archive.git}"
-: "${new_coq_opam_archive_git_branch:=master}"
+: "${new_coq_opam_archive_git_branch:=reference-arities}"
```
Comment (SkySkimmer) [inline dev/bench/bench.sh:75]: no such branch

Comment (mattam82) [reply dev/bench/bench.sh]: My bad again

**Comment by mattam82** [comment]
@coqbot bench

**Comment by mattam82** [comment]
@coqbot bench

**Comment by mattam82** [comment]
@ppedrot @SkySkimmer on current master this has a negligible effect, however on top of algebraic universes, math-classes goes from 149s to 123s just by forcing `equiv` and `le/lt` to unfold only when applied to at least 2 arguments (which is entirely backwards compatible). Wdyt? Should we pursue this?

**Comment by SkySkimmer** [comment]
This is on top of the algebraic univs PR right?

**Comment by mattam82** [comment]
Yep

**Pull request closed without merging**